### PR TITLE
perf(dashboard): implement widget item api v2

### PR DESCRIPTION
**Pull request by st3iny**
### ☑️ Resolves

* Ref https://github.com/nextcloud/server/pull/39937

### 🖼️ Screenshots

🏚️ Before | 🏡 After
---|---
B | A

### 🚧 Tasks

- [ ] ...

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not required
- [ ] 📘 API documentation in `docs/` has been updated or is not required
- [x] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

File: lib/Dashboard/TalkWidget.php [hunk 1]
```diff
@@ -112,7 +118,7 @@ public function getWidgetButtons(string $userId): array {
 		$buttons[] = new WidgetButton(
 			WidgetButton::TYPE_MORE,
 			$this->url->linkToRouteAbsolute('spreed.Page.index'),
-			$this->l10n->t('More unread mentions')
+			$this->l10n->t('More talk rooms')
```
Comment (nickvergessen) [inline lib/Dashboard/TalkWidget.php:121]:
```suggestion
			$this->l10n->t('More conversations')
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: lib/Dashboard/TalkWidget.php [hunk 2]
```diff
@@ -33,23 +34,27 @@
 use OCA\Talk\Room;
 use OCA\Talk\Service\AvatarService;
 use OCA\Talk\Service\ParticipantService;
+use OCA\Talk\TalkSession;
 use OCP\AppFramework\Utility\ITimeFactory;
 use OCP\Comments\IComment;
 use OCP\Dashboard\IAPIWidget;
+use OCP\Dashboard\IAPIWidgetV2;
```
Comment (nickvergessen) [inline lib/Dashboard/TalkWidget.php:41]:
```suggestion
```

File: lib/Dashboard/TalkWidget.php [hunk 3]
```diff
@@ -33,23 +34,27 @@
...
+use OCP\Dashboard\Model\WidgetItems;
 use OCP\Dashboard\Model\WidgetOptions;
 use OCP\IL10N;
 use OCP\IURLGenerator;
 use OCP\IUser;
 use OCP\IUserSession;
 use OCP\Util;
 
-class TalkWidget implements IAPIWidget, IIconWidget, IButtonWidget, IOptionWidget, IConditionalWidget {
+class TalkWidget implements IAPIWidget, IAPIWidgetV2, IIconWidget, IButtonWidget, IOptionWidget, IConditionalWidget, IReloadableWidget {
```
Comment (nickvergessen) [inline lib/Dashboard/TalkWidget.php:57]:
```suggestion
class TalkWidget implements IAPIWidget, IIconWidget, IButtonWidget, IOptionWidget, IConditionalWidget, IReloadableWidget {
```

File: lib/Dashboard/TalkWidget.php (hunk 1 above)
Comment (st3iny) [reply lib/Dashboard/TalkWidget.php]: Done

File: lib/Dashboard/TalkWidget.php [hunk 4]
```diff
@@ -170,6 +174,53 @@ public function getItems(string $userId, ?string $since = null, int $limit = 7):
 		return $result;
 	}
 
+	/**
+	 * @inheritDoc
+	 */
+	public function getItemsV2(string $userId, ?string $since = null, int $limit = 7): WidgetItems {
+		$sessionIds = $this->session->getAllActiveSessions();
+		$allRooms = $this->manager->getRoomsForUser($userId, $sessionIds, true);
```
Comment (nickvergessen) [inline lib/Dashboard/TalkWidget.php:182]: You are actually not using the sessions, so we can also skip fetching them.

```suggestion
		$allRooms = $this->manager->getRoomsForUser($userId, [], true);
```

File: lib/Dashboard/TalkWidget.php [hunk 5]
```diff
@@ -61,6 +66,7 @@ public function __construct(
 		protected ParticipantService $participantService,
 		protected MessageParser $messageParser,
 		protected ITimeFactory $timeFactory,
+		protected TalkSession $session,
```
Comment (st3iny) [inline lib/Dashboard/TalkWidget.php:69]:
```suggestion
```

**Comment by nickvergessen** [comment]
Arg, same problem with the search and smart picker now applies: #9867

**Review by nickvergessen — APPROVED** [review]

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**